exercises/src/main/scala/fpinscala/errorhandling/Either.scala

```diff
@@ -4,34 +4,57 @@ package fpinscala.errorhandling
 import scala.{Option => _, Either => _, Left => _, Right => _, _} // hide std library `Option` and `Either`, since we are writing our own in this chapter
 
 sealed trait Either[+E,+A] {
- def map[B](f: A => B): Either[E, B] = sys.error("todo")
-
- def flatMap[EE >: E, B](f: A => Either[EE, B]): Either[EE, B] = sys.error("todo")
-
- def orElse[EE >: E, B >: A](b: => Either[EE, B]): Either[EE, B] = sys.error("todo")
-
- def map2[EE >: E, B, C](b: Either[EE, B])(f: (A, B) => C): Either[EE, C] = sys.error("todo")
+ def map[B](f: A => B): Either[E, B] = this match {
+		case Left(e) => Left(e)
+		case Right(v) => Right(f(v))
+ }
+
+ def flatMap[EE >: E, B](f: A => Either[EE, B]): Either[EE, B] = this match {
+		case Left(e) => Left(e)
+		case Right(v) => f(v)
+ }
+
+ def orElse[EE >: E, B >: A](b: => Either[EE, B]): Either[EE, B] = this match {
+ 		case Left(e) => b
+ 		case Right(v) => Right(v)		
+ }
+
+ def map2[EE >: E, B, C](b: Either[EE, B])(f: (A, B) => C): Either[EE, C] = {
+ 		for {
+ 			aa <- this
+ 			bb <- b
+ 		} yield f(aa, bb)
+ }
 }
 case class Left[+E](get: E) extends Either[E,Nothing]
 case class Right[+A](get: A) extends Either[Nothing,A]
 
+case class Person(name: Name, age: Age)
+sealed class Name(v: String) 
+sealed class Age(v: Int) 
+
+
 object Either {
-  def traverse[E,A,B](es: List[A])(f: A => Either[E, B]): Either[E, List[B]] = sys.error("todo")
 
-  def sequence[E,A](es: List[Either[E,A]]): Either[E,List[A]] = sys.error("todo")
+	def traverse[E,A,B](es: List[A])(f: A => Either[E, B]): Either[E, List[B]] = 
+		es.foldRight[Either[E, List[B]]](Right(Nil))((h,t) => f(h).map2(t)(_ :: _))
+
+	def sequence[E,A](es: List[Either[E,A]]): Either[E,List[A]] = 
+		traverse(es)(x => x)
```

Comment:
`sequence` turns out to be surprisingly easy after `traverse`. As a convenience, Scala gives you a built-in function `identity` that is `x => x`.